deadlines/src/main/java/com/palantir/deadlines/Deadlines.java

```diff
@@ -0,0 +1,166 @@
+     *
+     * The actual deadline value encoded will be the minimum of:
+     *   - the providedDeadline parameter
+     *   - the value returned by {@link #getRemainingDeadline()}} if it exists
+     * This ensures that the deadline set for the request will be based on the remaining deadline from
+     * already-set internal state, or a smaller one if the caller chooses that.
+     *
+     * This function has no side-effects on the internal deadline state stored in a TraceLocal.
+     *
+     * @param proposedDeadline a proposed value for the deadline; the actual value used will be the minimum
+     * @param request the request object to write the encoding to
+     * @param adapter a {@link RequestEncodingAdapter} that handles writing the header value to the request object
+     * @throws DeadlineExpiredException if the actual deadline selected (per the above rules) has already expired
+     */
+    public static <T> void encodeToRequest(
+            Duration proposedDeadline, T request, RequestEncodingAdapter<? super T> adapter) {
+        Duration actualDeadline = proposedDeadline;
+        Optional<Duration> deadlineFromState = getRemainingDeadline();
+        if (deadlineFromState.isPresent() && deadlineFromState.get().compareTo(proposedDeadline) < 0) {
+            actualDeadline = deadlineFromState.get();
+        }
+
+        if (actualDeadline.isZero() || actualDeadline.isNegative()) {
+            throw new DeadlineExpiredException();
+        }
+
+        adapter.setHeader(request, DeadlinesHttpHeaders.EXPECT_WITHIN, durationToHeaderValue(actualDeadline));
+    }
+
+    /**
+     * Parse a deadline value from a request header and set the deadline state for the current trace.
+     *
+     * If the request object contains a deadline value in a header, this method will parse it and store
+     * the deadline value state internally in a TraceLocal, making it available to future calls to
+     * {@link #getRemainingDeadline()}} from threads participating in the current trace. The deadline value
+     * is read from a {@link DeadlinesHttpHeaders#EXPECT_WITHIN} header on the request object.
+     *
+     * If no such header exists, an {@link Optional#empty()} is returned.
+     *
+     * This function has side-effects on the internal deadline state stored in a TraceLocal; the state is
+     * set (or overwritten) based on the value of the deadline parsed from request headers.
+     *
+     * @param request the request object to read the deadline value from
+     * @param adapter a {@link RequestDecodingAdapter} that handles reading the header value from the request object
+     * @return the remaining deadline time parsed from the request headers, or {@link Optional#empty()} if
+     * no such deadline exists.
+     * @throws DeadlineExpiredException if the deadline parsed from the request is <= 0
+     */
+    public static <T> Optional<Duration> parseFromRequest(T request, RequestDecodingAdapter<? super T> adapter) {
```

Comment:
i'm not sure returning `Optional<Duration>` here is particularly useful. If a caller needs to read the deadline, they should maybe just call `getRemainingDeadline` instead.

Reply:
Agreed